File: 2 Definición y Características de la Recursión/PlayGround.java
public class PlayGround {
    public static void main(String[] args){

        Box box = new Box();
        box.value = 0;
        System.out.println(fibonacciRecursive(25, box));
        System.out.println("Invocations of fibonacciRecursive: " + String.valueOf(box.value));
    }

    private static int factorial (int n){
        int fminus1;
        System.out.println(n);
        if (n == 0) return 1;
        fminus1 = factorial(n-1);
        System.out.println(String.valueOf(n) + " " + String.valueOf(fminus1));
        return  n * fminus1;
    }

    private static String decimalToBinary (int n){
        if (n >= 2){
            return decimalToBinary(n/2) + String.valueOf(n%2);
        } else {
            return String.valueOf(n);
        }
    }

    private static int fibonacciRecursive (int n, Box box){
        // Save the number of access to the recursive fuction
        box.value++;


        if (n == 0){
            return 0;
        }
        if (n == 1){
            return 1;
        } else {
            return (fibonacciRecursive(n-1, box) + fibonacciRecursive(n-2, box));
        }
    }

    private static int fibonacci (int n){
        int fibonacci_a = 0;
        int fibonacci_b = 1;
        int fibonacci = 0;

        for (int i = 0; i < n; i++){
            fibonacci = fibonacci_b + fibonacci_a;
            fibonacci_b = fibonacci_a;
            fibonacci_a = fibonacci;
        }

        return fibonacci;
    }

    private static class Box{
        public int value;
    }
}

File: 2 Definición y Características de la Recursión/Playground.kt
fun main (){
    // External variable to count the nummber of call to
    // the factorial recursive fuction
    var calls = Box(0);
    println(fibonacciRecursive(3, calls));
    println("Number of call to fibonacci recursive function: " + calls.value.toString());
}

fun factorial (n: Int): Int{
    var fminus1: Int;
    println(n);
    if (n == 0) return 1;
    fminus1 = factorial(n-1);
    println(n.toString() + " " + fminus1.toString());
    return (n * fminus1);
}

fun decimalToBinary (decimal: Int): String{
    if (decimal >= 2){
        return decimalToBinary(decimal/2) + (decimal%2).toString();
    } else {
        return decimal.toString();
    }
}

fun fibonacciRecursive (n: Int, calls: Box): Int{
    // Register another call to the recursive funcion.
    calls.value++;
    when {
        (n == 0) -> {
            return 0;
        }

        (n == 1) -> {
            return 1;
        }

        else -> {
            return (fibonacciRecursive(n-1, calls) + fibonacciRecursive(n-2, calls));
        }
    }
}

fun fibonacci (n: Int): Int{
    var previousA = 0;
    var previousB = 1;
    var fibonacci = 0;

    for (i in 0 .. n-1){
        fibonacci = previousB + previousA;
        previousB = previousA;
        previousA = fibonacci;
    }

    return fibonacci;
}

// An integer container that we can 
// use like pointer to int
class Box (value: Int){
    var value: Int = value;
}